test/Test/Database/LSMTree/Internal/Merge.hs

```diff
@@ -0,0 +1,149 @@
+
+    lhsKOpsFile <- FS.hGetAll fs (Run.lsmRunKOpsFile lhs)
+    lhsBlobFile <- FS.hGetAll fs (Run.lsmRunBlobFile lhs)
+    rhsKOpsFile <- FS.hGetAll fs (Run.lsmRunKOpsFile rhs)
+    rhsBlobFile <- FS.hGetAll fs (Run.lsmRunBlobFile rhs)
+
+    -- cleanup
+    traverse_ (Run.removeReference fs) runs
+    Run.removeReference fs lhs
+    Run.removeReference fs rhs
+
+    return $ stats $
+           Run.lsmRunNumEntries lhs === Run.lsmRunNumEntries rhs
+      .&&. Run.lsmRunFilter lhs === Run.lsmRunFilter rhs
+      .&&. Run.lsmRunIndex lhs === Run.lsmRunIndex rhs
+      .&&. lhsKOpsFile === rhsKOpsFile
+      .&&. lhsBlobFile === rhsBlobFile
+  where
+    flush n = Run.fromWriteBuffer fs (Run.RunFsPaths n)
+
+    stats = tabulate "value size" (map (showPowersOf10 . sizeofValue) vals)
+          . label (if any isLargeKOp kops then "has large k/op" else "no large k/op")
+          . label ("number of runs: " <> showPowersOf10 (length wbs))
+    kops = foldMap WB.content wbs
+    vals = concatMap (bifoldMap pure mempty . snd) kops
+
+-- | After merging for a few steps, we can prematurely abort the merge, which
+-- should clean up properly.
+prop_CloseMerge ::
+     FS.HasFS IO h -> FS.HasBufFS IO h ->
+     StepSize ->
+     [WriteBuffer KeyForCompactIndex SerialisedValue SerialisedBlob] ->
+     IO Property
+prop_CloseMerge fs bfs (Positive stepSize) wbs = do
+    let path0 = Run.RunFsPaths 0
+    runs <- sequenceA $ zipWith flush [10..] wbs
+    mergeToClose <- makeInProgressMerge path0 runs
+    traverse_ (Merge.close fs) mergeToClose
+
+    filesExist <- traverse (FS.doesFileExist fs) (Run.runFsPaths path0)
+
+    -- cleanup
+    traverse_ (Run.removeReference fs) runs
+
+    return $
+      counterexample ("run files exist: " <> show filesExist) $
+        isJust mergeToClose ==> all not filesExist
+  where
+    flush n = Run.fromWriteBuffer fs (Run.RunFsPaths n)
+
+    makeInProgressMerge path runs =
+      Merge.new fs bfs path runs >>= \case
+        Nothing -> return Nothing  -- not in progress
+        Just merge -> do
+          -- just do a few steps once, ideally not completing the merge
+          Merge.steps fs bfs mergeValues merge stepSize >>= \case
+            Merge.MergeComplete run -> do
+              Run.removeReference fs run-- run not needed, close
+              return Nothing  -- not in progress
+            Merge.MergeInProgress ->
+              return (Just merge)
+
+{-------------------------------------------------------------------------------
+  Utilities
+-------------------------------------------------------------------------------}
+
+type StepSize = Positive Int
+
+mergeRuns ::
+     FS.HasFS IO h -> FS.HasBufFS IO h ->
+     Int ->
+     [Run.Run (FS.Handle h)] ->
+     StepSize ->
+     IO (Run.Run (FS.Handle h))
+mergeRuns fs bfs n runs (Positive stepSize) = do
+    Merge.new fs bfs (Run.RunFsPaths n) runs >>= \case
+      Nothing -> Run.fromWriteBuffer fs (Run.RunFsPaths n) WB.empty
```

Comment:
As long as the input runs are non-empty, we'll get a merge. On the last level, this might still result in an empty run, although that is extremely unlikely.

Reply:
Also, thanks for mentioning "last level". Currently there's no support for doing last level merge, for which I should add an argument.

Reply:
Some of these edge cases might not actually be reachable, but it's probably a useful invariant to have that none of the runs can be empty except for the last level

Reply:
I added support for last level merges, to be specified when creating the merge.